src/client/pythonEnvironments/base/locators/common/wrappingLocator.ts

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Event } from 'vscode';
+import { IDisposable } from '../../../../common/utils/resourceLifecycle';
+import { PythonEnvInfo } from '../../info';
+import {
+    ILocator, IPythonEnvsIterator, PythonLocatorQuery,
+} from '../../locator';
+import { PythonEnvsChangedEvent, PythonEnvsWatcher } from '../../watcher';
+import { LazyResourceBasedLocator } from './resourceBasedLocator';
+
+export type GetLocatorFunc = () => Promise<ILocator & Partial<IDisposable>>;
+
+/**
+ * A locator that wraps another.
+ *
+ * This facilitates isolating the wrapped locator.
+ */
+export class LazyWrappingLocator extends LazyResourceBasedLocator {
+    public readonly onChanged: Event<PythonEnvsChangedEvent>;
+
+    private readonly watcher = new PythonEnvsWatcher();
+
+    private wrapped?: ILocator;
+
+    constructor(
+        private readonly getLocator: GetLocatorFunc,
+    ) {
+        super();
+        this.onChanged = this.watcher.onChanged;
+    }
+
+    protected async* doIterEnvs(query?: PythonLocatorQuery): IPythonEnvsIterator {
+        yield* this.wrapped!.iterEnvs(query);
+    }
+
+    protected async doResolveEnv(env: string | PythonEnvInfo): Promise<PythonEnvInfo | undefined> {
+        return this.wrapped!.resolveEnv(env);
+    }
+
+    protected async initResources(): Promise<IDisposable[]> {
+        const locator = await this.getLocator();
+        this.wrapped = locator;
+        if (locator.dispose !== undefined) {
+            return [locator as IDisposable];
+        }
+        return [];
```

Comment:
```suggestion
        if (locator.dispose !== undefined) {
            this.disposables(locator as IDisposable);
        }
```
We can simplify the return type to `Promise<void>` if we do this instead.

Reply:
I've considered both approaches.  Returning the disposables makes it more clear what is expected of the subclass.  It makes it less likely that the implementation will lose track of disposables.  That said, the difference isn't much.  As long as we don't lose track of disposables then I'm fine with just returning `void`.